src/components/VmActions/AutoSelectConsole.js

```diff
@@ -0,0 +1,104 @@
+import React, { useCallback, useEffect, useState } from 'react'
+import { connect } from 'react-redux'
+import PropTypes from 'prop-types'
+import { downloadVmConsoles } from '_/actions'
+import ConsoleConfirmationModal from '_/components/VmActions/ConsoleConfirmationModal'
+import {
+  EMPTY_CONSOLES_LIST,
+  NO_DEFAULT_CONSOLE,
+  DISPATCH_STATUSES,
+} from '_/constants'
+
+const {
+  DO_NOT_DISPATCH,
+  SHOULD_DISPATCH,
+  DISPATCHED,
+} = DISPATCH_STATUSES
+
+const DEFAULT_CONSOLE_PROTOCOL_ERRORS_LIST = [
+  NO_DEFAULT_CONSOLE,
+  EMPTY_CONSOLES_LIST,
+]
+
+const AutoSelectConsole = ({
+  vm,
+  show,
+  isNoVnc,
+  fetchConsoles,
+  id,
+  onClose,
+  defaultConsole,
+}) => {
+  const [dispatchStatus, setDispatchStatus] = useState(DO_NOT_DISPATCH)
+
+  const handleClose = useCallback(() => {
+    setDispatchStatus(DO_NOT_DISPATCH)
+    onClose()
+  }, [])
+
+  useEffect(() => {
+    if (show) {
+      setDispatchStatus(SHOULD_DISPATCH)
+      fetchConsoles(vm)
+    }
+  }, [show, fetchConsoles])
+
+  useEffect(() => {
+    if (show && dispatchStatus === SHOULD_DISPATCH) {
+      if (!defaultConsole || defaultConsole === 'undefined') {
+        return
+      }
+
+      if (DEFAULT_CONSOLE_PROTOCOL_ERRORS_LIST.includes(defaultConsole)) {
+        handleClose()
+      } else {
+        setDispatchStatus(DISPATCHED)
+      }
+    }
+  }, [defaultConsole, dispatchStatus, show, handleClose])
+
+  const isProtocolSelected = show && dispatchStatus === DISPATCHED
+  return (
+    <div>
+      {isProtocolSelected &&
+      <ConsoleConfirmationModal
+        consoleId={defaultConsole.id}
+        vm={vm}
+        show
+        isNoVNC={defaultConsole.protocol === 'vnc' ? isNoVnc : false}
+        onClose={() => {
+          if (dispatchStatus === DISPATCHED) {
+            handleClose()
+          }
+        }}
+        id={id}
+      />
+      }
+    </div>
+  )
+}
```

Comment:
I think you can get rid of component state use completely by just doing:

```jsx
}) => {
  useEffect(() => {
    if (show && defaultConsole === 'undefined') {
      fetchConsoles()
    }
  }, [show, defaultConsole])

  if (!show || defaultConsole === 'undefined' || defaultConsole === EMPTY_CONSOLES_LIST) {
    return null
  }

  return (
    <ConsoleConfirmationModal
      show
      vm={vm}
      consoleId={defaultConsole.id}
      isNoVNC={defaultConsole.protocol === 'vnc' ? isNoVnc : false}
      onClose={onClose}
      id={id}
    />
  )
}
```

Note 1: If the console is already fetched for the VM, it won't be fetched again.

Note 2: The `defaultConsole === 'undefined'` value is set in the base VM transform.  You can change it to whatever constant makes sense.  

In the case of a headless VM, clicking the action button will set `show=true` and then with `defaultConsole === 'undefined'` will hit the effect and call the fetch saga.  But, the saga would then set `deafulConsole` to `EMPTY_CONSOLES_LIST`.  The fetch won't be triggered again.  Only an error toast on the first click will be displayed.  Subsequent clicks will have no effect.

Perhaps if a VM is known to be headless, the console button can be removed from the action card or otherwise disabled to avoid confusion.  @sgratch, @rszwajko - Thoughts on multiple clicks to the console button for a headless VM?

Reply:
@sjd78
I got rid of the component's state but I made few small changes to integrate between the components because `ConsoleConfirmationModal` is class component and `AutoSelectConsole` is functional ConsoleConfirmationModal sometimes triggers the onClose with no reason and then the console wasn't fetched as expected.

Reply:
@sjd78 @bamsalem 

> Perhaps if a VM is known to be headless, the console button can be removed from the action card or otherwise disabled to avoid confusion. @sgratch, @rszwajko - Thoughts on multiple clicks to the console button for a headless VM?

I think it might be confusing for the user if the second click will disable the console button till next hard refresh or console configuration changes on backend. I think it's better to leave it as is to keep the ui simple and persistent.

Reply:
> I think you can get rid of component state use completely

yep. This is the right direction.

> Perhaps if a VM is known to be headless, the console button can be removed from the action card or otherwise disabled to avoid confusion. @sgratch, @rszwajko - Thoughts on multiple clicks to the console button for a headless VM?

I would disable the action (make the button grey) to clearly indicate that this option is not available. Active buttons encourage users to click & try them i.e. users may try it few times hoping it will finally work.

Reply:
@sjd78 @rszwajko 
![image](https://user-images.githubusercontent.com/64131213/108827491-597cdc80-75ce-11eb-964c-eaa8bb524be4.png)

>I would disable the action (make the button grey) to clearly indicate that this option is not available. Active buttons encourage users to click & try them i.e. users may try it few times hoping it will finally work.

We can't disable the button without disabling the dropdown menu (with the current component, PF component).
and I'm not sure it's worth to create our own component for that, but we must add another notifier, maybe keep showing the error on every click...
WDYT?

Reply:
@bamsalem for headless:
I suggest reverting back for keeping the headless error per every click.
 We can change the menu items order so that "console" menu item will not be the default till next refresh (the default/first item will be "Remote Desktop" for win based vms and "Suspend" for others).

Reply:
> @bamsalem for headless:
> I suggest reverting back for keeping the headless error per every click.

Done.

> We can change the menu items order so that "console" menu item will not be the default till next refresh (the default/first item will be "Remote Desktop" for win based vms and "Suspend" for others).

Regarding changing the order, it's more complex than it looks, because it's a PF component when we change the order it's change the confirmation modal and than it display non related notification, for example "Are you sure you want to Suspend the VM?", to handle that we need to implement our own component.

Reply:
@bamsalem so we can go back to original behavior of just hidden/dismiss the "console" menu item till next refresh.